### PR TITLE
[Draft]integrate triton into ort

### PR DESCRIPTION
**Pull request by kailums**
### Description

Here we found in some scenarios, triton written kernel is more performant than CK or other kernels. So we want onnxruntime to use these triton written kernels.

This PR implements a framework that supports write and compile a triton kernel, and call it from onnxruntime.

#### Write and compile triton kernel
We use softmax as example. We have implemented a softmax kernel using triton at `onnxruntime/python/tools/kernel_explorer/kernels/rocm/triton/softmax.py`

```
@triton.jit
def softmax_kernel(
    output_ptr, input_ptr, input_row_stride, output_row_stride, n_cols,
    BLOCK_SIZE: tl.constexpr
):
    # softmax implementings
    ....
    ....
```
For each kernel, there needs to add description info:
```
# kernel dtype and BLOCK_SIZE to generate.
dtypes = ['fp32', 'fp16']
blocks = [1024, 2048, 4096, 8192, 16384]
name_pattern = 'softmax_{}_{}'
sig_pattern = '*{},*{},i32,i32,i32'

# this function returns a description table that contains function name and metadata for launching kernels
def GetSoftmaxFunctionTable():
    ......
    ......
```

And then add `GetSoftmaxFunctionTable` function into  `onnxruntime/python/tools/kernel_explorer/kernels/rocm/triton/compile_triton.py` to register this kernel.
```
kernel_table = [
    GetSoftmaxFunctionTable,  # softmax
]
```
Then run `python compile_triton.py` and get the compiled kernel files in `libs`:
```
-rw-r--r-- 1 root root  1423 Apr 11 05:47 meta.json
-rw-r--r-- 1 root root  9720 Apr 11 05:47 softmax_fp16_1024.hasco
-rw-r--r-- 1 root root 17872 Apr 10 09:17 softmax_fp16_16384.hasco
-rw-r--r-- 1 root root  9720 Apr 11 05:47 softmax_fp16_2048.hasco
-rw-r--r-- 1 root root  9720 Apr 11 05:47 softmax_fp16_4096.hasco
-rw-r--r-- 1 root root 13776 Apr 10 09:17 softmax_fp16_8192.hasco
-rw-r--r-- 1 root root  9720 Apr 11 05:47 softmax_fp32_1024.hasco
-rw-r--r-- 1 root root 17872 Apr 10 09:17 softmax_fp32_16384.hasco
-rw-r--r-- 1 root root  9720 Apr 11 05:47 softmax_fp32_2048.hasco
-rw-r--r-- 1 root root  9720 Apr 11 05:47 softmax_fp32_4096.hasco
-rw-r--r-- 1 root root 13776 Apr 10 09:17 softmax_fp32_8192.hasco
```
The `meta.json` file contains metadata used for launching kernel, and `*.hsaco` files are compiled triton kernel for AMD GPU. it will be `*.cubin` for Nvidia GPU.

#### onnxruntime c++ code moidfication
To use there kernels in onnxruntime, we need to implement a call stub.
Here we implement a tunable op for softmax in `onnxruntime/core/providers/rocm/math/softmax_triton.cuh`
```
template <typename T, typename OutputT>
Status SoftmaxTritonOp(const SoftmaxParams<T, OutputT>* params) {
  if (params->is_log_softmax) {
    TUNABLE_OP_RETURN_UNSUPPORTED_ARGUMENT_IF(true, "softmax triton not support log-softmax");
  }
  auto fname = GetSoftmaxTritonName<T>(params->softmax_elements);

  // construct args for launch kernel
  struct {
    void *out;
    const void *in;
    int in_stride;
    int out_stride;
    int n_cols;
  } args = {(void*)params->output, (const void*)params->input, params->input_stride, params->output_stride, params->softmax_elements};

  // grid dim is (batch_count, 1, 1)
  return LaunchTritonKernel(params->stream, fname, params->batch_count, 1, 1, &args, sizeof(args));
}
```
#### runtime
The onnxruntime needs to load compiled triton kernel when executing operators.
Here is an environment variable `ORT_TRITON_LIB_PATH` needs to be set to the path of compiled trtion kernels, and should use absolute path, for example:
`export ORT_TRITON_LIB_PATH=/ws/work/onnxruntime/onnxruntime/python/tools/kernel_explorer/kernels/rocm/triton/libs`

#### Test
Using kernel_explorer, we can test this softmax kernel.
```
export KERNEL_EXPLORER_BUILD_DIR=/ws/code/onnxruntime/build_rocm/Debug
export ORT_TRITON_LIB_PATH=/ws/code/onnxruntime/onnxruntime/python/tools/kernel_explorer/kernels/rocm/triton/libs

python onnxruntime/python/tools/kernel_explorer/kernels/softmax_test.py
```
and can see the results, here TunableOp select SoftmaxTriton, it's better than others.
```
SoftmaxTunable               float16 batch_count=1    softmax_elements=2048 is_log_softmax=0   7.05 us, 1.16 GB/s
SoftmaxTriton                float16 batch_count=1    softmax_elements=2048 is_log_softmax=0   5.17 us, 1.59 GB/s
SoftmaxBlockwise_2           float16 batch_count=1    softmax_elements=2048 is_log_softmax=0   9.48 us, 0.86 GB/s
....
....
```

### CODE REVIEW

File: build_rocm.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  --rocm_home ${ROCM_HOME} \
+  --nccl_home ${ROCM_HOME}\
+  --update \
+  --build_dir ${BUILD_DIR} \
+  --build \
+  --parallel \
+  --build_wheel \
+  --skip_tests \
+  --skip_submodule_sync \
+  --allow_running_as_root
```
Comment (snnn) [inline build_rocm.sh:33]: Why “  --allow_running_as_root”？ Who would run this script with root?

Comment (kailums) [reply build_rocm.sh:33]: Thank for reviewing. I was running this build script in a docker, and it needs to add this flag.
Seems this script is not useful for others, may be I should remove it from this PR.

**Comment by wschin** [comment]
I only see very limited documentation in the codebase. Can you please explain those newly added components in detail?

File: onnxruntime/python/tools/kernel_explorer/kernels/rocm/softmax.cu [hunk 2]
```diff
@@ -171,6 +172,37 @@ class CKSoftmax : public IKernelExplorer {
 };
 #endif  // USE_COMPOSABLE_KERNEL
 
+#ifdef USE_TRITON_KERNEL
+template <typename T>
+class SoftmaxTriton : public IKernelExplorer {
```
Comment (wschin) [inline onnxruntime/python/tools/kernel_explorer/kernels/rocm/softmax.cu:177]: Humm, this doesn't look right. I remember we have a `custom op` mechanism that can register arbitrary kernels. Any concerns on using `custom op`?

Comment (codego7250) [reply onnxruntime/python/tools/kernel_explorer/kernels/rocm/softmax.cu:177]: This is to take advantage of the tuned kernel for AMD Rocm EP.
IkernelExplore will help to select the best performant Kernels from various sources, including triton kernels.

**Comment by kailums** [comment]
> I only see very limited documentation in the codebase. Can you please explain those newly added components in detail?

yeah, The main component I added is a loader and launcher API for triton compiled kernel. That kernel is a hsaco or cubin format file.  And the other part is an implementation for softmax using triton compiled kernel.

For the ops like softmax, I write a triton script and compile script that can compile triton code into hsaco file offline. 

When onnxruntime is doing online inference, I provide an env variable to the path of hsaco file, onnxruntime use the loader to load those hsaco files and register kernel functions into a global registry. Then if it runs into softmax op, it can search the registry and launch the triton compiled softmax kernel using launcher API.

For performance reason, we not directly implement a new softmax Op for triton, we use a TunableOP, that contains multi implementations for an op. For softmax op, we have a handwritten kernel, and a Composable Kernel implementation, and trtion written kernel. The tunableOP will compare these implementations, and select the most efficient one.

**Pull request closed without merging**